### PR TITLE
feat: better configuration

### PR DESCRIPTION
**Pull request by marcus8448**
# Changes

## Basic stuff
- Rewrote the machine tab/panel screen
- Rewrote how side configurations are handled
  - Can now be filtered per slot and slot type
- Rewrote how dynamic machine models are rendered (allowing for items to be dynamically rendered too!)
- Made a bunch of package names singular to follow fabric conventions
- **Changed the way you add tanks and inventory slots to machines**
- Updated GH actions workflows
- Bumped loom to 0.7 and gradle to 7
- Added WTHIT (WAILA) support to GC:R
- Bumped a bunch of dependencies
- The logger now prepends "[Galacticraft] " to all messages
- Fixes GC player inventory oxygen overlay
- Fixes oxygen HUD
- Makes coal generator glow like a furnace
- Fixes a potential exploit in machine config networking code
- Rewrote wires for the fourth time
- Squashed bugs related to attributes
- You can directly insert/extract fluids from tanks
- Slots/tanks get highlighted when configuration face is hovered and filtered
- Machine packets no longer redundantly send `BlockPos`es as the server already knows what SH is open
- More bugs

### Videos of things

Direct Tank Insertion: https://youtu.be/YpSMbKO5dl4
Machine I/O v2: https://youtu.be/zIZ9hq0vvcM

## Technical Changes

`ConfigurableMachineBlockEntity` -> `MachineBlockEntity`
`ConfigurableMachineBlock` -> `MachineBlock`
`SideConfig` -> `AutomationType`

### `SlotType`s

A new system that groups slots together for I/O filtering.
Must be registered in `SlotType#SLOT_TYPES` 
Contains an `AutomationType`, name, id, and colour (not locked to MC `Formatting`)

### You must add inventory slots in the `MachineBlockEntity#createInventory` method

I've rewritten the inventory systems to enforce slot filters and to remove the need to go to like 3 classes to add a slot and 4 methods to filter player and automatic insertion/extraction.
```java
    @Override
    protected MachineItemInv.Builder createInventory(MachineItemInv.Builder builder) {
        builder.addSlot(CHARGE_SLOT, SlotType.CHARGE, EnergyUtils.IS_EXTRACTABLE, 8, 8);
        builder.addSlot(FUEL_SLOT, SlotType.COAL, stack -> FUEL_MAP.containsKey(stack.getItem()), 8, 74);
//        builder.addSlot(index, type, filter, slotX, slotY);
        return builder;
    }
```

### You must add fluid tanks in the `MachineBlockEntity#createFluidInv` method

No need to manually add widgets to render a tank.
Automatically synced via SH
Simplifies filtering and I/O configurations
```java
    @Override
    protected MachineFluidInv.Builder createFluidInv(MachineFluidInv.Builder builder) {
        builder.addTank(OIL_TANK, SlotType.OIL_IN, Constants.Filter.OIL, 122, 28, 1);
        builder.addTank(FUEL_TANK, SlotType.FUEL_OUT, Constants.Filter.FUEL, 152, 28, 1);
//        builder.addTank(index, type, filter, tankX, tankY, tankScale);
        return builder;
    }
```

### Machine model rendering changes

You must manually state texture dependencies for machine models, as they are no longer loaded by item models (since they are dynamic now). See `MachineBakedModel#TEXTURE_DEPENDENCIES`

### CODE REVIEW

**Comment by vos6434** [comment]
Always loved more more bugs 😆

**Review by JoeZwet — CHANGES_REQUESTED** [review]
some things still aren't renamed.

crash from trying to load moon village: https://pastebin.com/EiDJ28ZZ

File: src/main/java/dev/galacticraft/mod/item/OxygenTankItem.java [hunk 1]
```diff
@@ -56,51 +55,68 @@
...
     }
 
+    @Override
+    public int getEnchantability() {
+        return -1;
+    }
+
+    @Override
+    public boolean isEnchantable(ItemStack stack) {
+        return false;
```
[the remark this answers is not in the document]
Comment (JoeZwet) [reply src/main/java/dev/galacticraft/mod/item/OxygenTankItem.java:103]: it does not

File: src/main/java/dev/galacticraft/mod/item/GalacticraftArmorMaterial.java [hunk 2]
```diff
@@ -36,21 +36,45 @@
...
+public enum GalacticraftArmorMaterial implements ArmorMaterial {
+    SENSOR_GLASSES("sensor_glasses",
+            0,
+            new int[]{0, 0, 0, 0},
+            0,
+            SoundEvents.ITEM_ARMOR_EQUIP_IRON,
+            0.0f,
+            () -> Ingredient.ofItems(GalacticraftItems.METEORIC_IRON_INGOT),
+            0.0f
+    ), // TODO: add actual functionality
```
[the remark this answers is not in the document]
Comment (marcus8448) [reply src/main/java/dev/galacticraft/mod/item/GalacticraftArmorMaterial.java:48]: That was there previously. I think #111 removes it.

File: src/main/java/dev/galacticraft/mod/client/gui/widget/machine/OxygenTankWidget.java [hunk 3]
```diff
@@ -80,15 +76,20 @@ public void drawTexture(MatrixStack matrices, int x, int y, int u, int v, int wi
...
+            lines.add(new TranslatableText("ui.galacticraft.machine.current_oxygen", new LiteralText(Screen.hasShiftDown() ? getView().get().amount().toString() + "B" : (getView().get().amount().asInt(1000, RoundingMode.HALF_DOWN) + "mB")).setStyle(Constants.Text.BLUE_STYLE)).setStyle(Constants.Text.GOLD_STYLE));
+            lines.add(new TranslatableText("ui.galacticraft.machine.max_oxygen", new LiteralText(String.valueOf((int)(OxygenCollectorBlockEntity.MAX_OXYGEN.asInt(1000, RoundingMode.HALF_DOWN)))).setStyle(Constants.Text.BLUE_STYLE)).setStyle(Constants.Text.RED_STYLE));
 
             this.client.currentScreen.renderTooltip(matrices, lines, mouseX, mouseY);
         }
     }
 
+    @Override
+    public void renderOutline(TextColor color) {
+//todo
```
[the remark this answers is not in the document]
Comment (marcus8448) [reply src/main/java/dev/galacticraft/mod/client/gui/widget/machine/OxygenTankWidget.java:90]: Originally, I was going to implement the systems in a totally different way - I'll remove the file.

File: src/main/java/dev/galacticraft/mod/Constants.java [hunk 4]
```diff
@@ -640,17 +616,55 @@ static String getRaw(String path) {
...
         String BLOCK_ENTITY_TAG = "BlockEntityTag";
+        String NO_DROP = "NoDrop";
+        String TOTAL_OXYGEN = "TotalOxygen";
+        String OXYGEN = "Oxygen";
+        String OWNER = "Owner";
+        String PROGRESS = "Progress";
+        String SIZE = "Size";
+        String MAX_SIZE = "MaxSize";
+        String GCR_DATA = "GCR_DATA";
+        String FUEL_TIME = "FuelTime";
```
[the remark this answers is not in the document]
Comment (marcus8448) [reply src/main/java/dev/galacticraft/mod/Constants.java:641]: Resolved in ae90810f21a890433fe05db89270368109a1dff7.

File: src/main/java/dev/galacticraft/mod/compat/hook/ModMenuApiImpl.java [hunk 5]
```diff
@@ -20,11 +20,11 @@
  * SOFTWARE.
  */
 
-package dev.galacticraft.mod.hooks;
+package dev.galacticraft.mod.compat.hook;
```
[the remark this answers is not in the document]
Comment (marcus8448) [reply src/main/java/dev/galacticraft/mod/compat/hook/ModMenuApiImpl.java]: Changed to `dev.galacticraft.mod.compat.modmenu` in ae90810f21a890433fe05db89270368109a1dff7.

**Review by JoeZwet — APPROVED** [review]

File: src/main/java/dev/galacticraft/mod/Galacticraft.java [hunk 6]
```diff
@@ -65,50 +66,49 @@
...
         GalacticraftDimensions.register();
         GalacticraftScreenHandlerTypes.register();
-        GalacticraftParticles.register();
-        GalacticraftCommands.register();
-        GalacticraftS2CPacketReceivers.register();
-        GalacticraftSounds.register();
-        GalacticraftBannerPatterns.register();
+        GalacticraftParticle.register();
+        GalacticraftCommand.register();
+        GalacticraftServerPacketReceivers.register();
```
Comment (craftablescience) [inline src/main/java/dev/galacticraft/mod/Galacticraft.java:97]: Would this be `GalacticraftServerPacketReceiver` to go with the singular naming scheme?

**Review by craftablescience — CHANGES_REQUESTED** [review]
This is the only thing I noticed.

File: src/main/java/dev/galacticraft/mod/Galacticraft.java (hunk 6 above)
Comment (craftablescience) [reply src/main/java/dev/galacticraft/mod/Galacticraft.java:97]: Actually now I see a few more plural things:
image
image
image
image
image

**Review by craftablescience — APPROVED** [review]
👍

**Pull request merged**